pkg/apis/apidiscovery/types.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apidiscovery
+
+import (
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// APIGroupDiscoveryList is a resource containing a list of APIGroupDiscovery.
+// This is one of the types able to be returned from the /api and /apis endpoint and contains an aggregated
+// list of API resources (built-ins, Custom Resource Definitions, resources from aggregated servers)
+// that a cluster supports.
+type APIGroupDiscoveryList struct {
+	v1.TypeMeta
+	// ResourceVersion will not be set, because this does not have a replayable ordering among multiple apiservers.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	v1.ListMeta
+
+	// items is the list of groups for discovery.
```

Comment:
priority order matters here as well, right?

Reply:
Priority order matters for versions within a group. This struct is just the list of groups that would probably be sorted in alphabetical order, I don't know if that warrants an extra comment?

Reply:
groups are listed in priority order as well. That tells kubectl to prefer `deployments` in the `apps` API group over `deployments` in an `example.com` CRD-based group. See the docs on the `groupPriorityMinimum` field in APIService

Reply:
The relative positioning of groups mattering is another obstacle to making this a watchable API where updates sent only a single APIGroupDiscovery entry

Reply:
Thanks for the clarification, updated.